File: rln/ffi_nim_examples/partial_proof.nim
include "common"

proc main() =
  var rlnInstance = initRLN(false)
  if rlnInstance.isNil:
    return

  var member = createMember()

  let merkleProof = registerMember(rlnInstance, member.rateCommitment)
  if merkleProof.isNil:
    return

  let externalNullifier = computeExternalNullifier()

  echo "\nHashing signal"
  var signal: array[32, uint8] = [1'u8, 2, 3, 4, 5, 6, 7, 8, 9, 10, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  let x = hashSignal(signal)
  printFr("x", x)

  echo "\nCreating message id"
  let messageId = ffi_uint_to_fr(0'u32)
  printFr("message id", messageId)

  echo "\nCreating RLN witness"
  let witnessResult = createWitness(member, merkleProof, messageId, x,
      externalNullifier)
  if witnessResult.ok.isNil:
    stderr.writeLine("Witness creation error: " & asString(witnessResult.err))
    ffi_c_string_free(witnessResult.err)
    return
  var witness = witnessResult.ok
  echo "  - RLN witness created successfully"

  echo "\nCreating partial witness from witness fields"
  let witnessIdentitySecret = ffi_rln_witness_input_get_identity_secret(witness)
  let witnessUserMessageLimit = ffi_rln_witness_input_get_user_message_limit(witness)
  let witnessMerkleProof = ffi_rln_witness_input_get_merkle_proof(witness)
  let partialWitnessResult = ffi_rln_partial_witness_input_new(
      witnessIdentitySecret, witnessUserMessageLimit, witnessMerkleProof)
  ffi_secret_fr_free(witnessIdentitySecret)
  ffi_fr_free(witnessUserMessageLimit)
  ffi_rln_merkle_proof_free(witnessMerkleProof)
  if partialWitnessResult.ok.isNil:
    stderr.writeLine("Partial witness creation error: " & asString(
        partialWitnessResult.err))
    ffi_c_string_free(partialWitnessResult.err)
    return
  var partialWitness = partialWitnessResult.ok
  echo "  - partial witness created successfully"

  echo "\nGenerating partial ZK proof"
  let partialProofResult = ffi_rln_generate_partial_proof(rlnInstance,
      partialWitness)
  if partialProofResult.ok.isNil:
    stderr.writeLine("Partial proof generation error: " & asString(
        partialProofResult.err))
    ffi_c_string_free(partialProofResult.err)
    return
  var partialProof = partialProofResult.ok
  echo "  - partial proof generated successfully"

  echo "\nFinishing proof with full witness"
  let fullProofResult = ffi_rln_finish_proof(rlnInstance,
      partialProof, witness)
  if fullProofResult.ok.isNil:
    stderr.writeLine("Finish proof error: " & asString(fullProofResult.err))
    ffi_c_string_free(fullProofResult.err)
    return
  var fullProof = fullProofResult.ok
  echo "  - partial proof finished successfully"

  echo "\nVerifying full proof"
  let verifyFullResult = verifyStatefulProof(rlnInstance, fullProof, x)
  if verifyFullResult.err.dataPtr != nil:
    stderr.writeLine("Full proof verification error: " & asString(
        verifyFullResult.err))
    ffi_c_string_free(verifyFullResult.err)
    return
  if verifyFullResult.ok:
    echo "  - full proof verified successfully"
  else:
    echo "Full proof verification failed"
    return

  ffi_rln_proof_free(fullProof)
  ffi_rln_partial_proof_free(partialProof)
  ffi_rln_partial_witness_input_free(partialWitness)
  ffi_rln_witness_input_free(witness)
  ffi_fr_free(messageId)
  ffi_fr_free(x)
  ffi_fr_free(externalNullifier)
  ffi_rln_merkle_proof_free(merkleProof)
  memberFree(member)
  ffi_rln_free(rlnInstance)

main()

File: rln/ffi_c_examples/partial_proof.c
#include <stdio.h>
#include <stdlib.h>

#include "common.c"

int main(void)
{
    RLN *rln_instance = init_rln(false);
    if (!rln_instance)
    {
        return EXIT_FAILURE;
    }

    Member member;
    create_member(&member);

    MerkleProof *merkle_proof = register_member(&rln_instance, member.rate_commitment);
    if (!merkle_proof)
    {
        return EXIT_FAILURE;
    }

    CFr *external_nullifier = compute_external_nullifier();

    printf("\nHashing signal\n");
    uint8_t signal[32] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 0, 0, 0, 0, 0, 0,
                          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    CFr *x = hash_signal(signal);
    print_cfr("x", x);

    printf("\nCreating message id\n");
    CFr *message_id = ffi_uint_to_cfr(0);
    print_cfr("message id", message_id);

    printf("\nCreating RLN witness\n");
    WitnessResult witness_result =
        create_witness(&member, merkle_proof, message_id, x, external_nullifier);
    if (!witness_result.ok)
    {
        fprintf(stderr, "Witness creation error: %s\n", witness_result.err.ptr);
        ffi_c_string_free(witness_result.err);
        return EXIT_FAILURE;
    }
    Witness *witness = witness_result.ok;
    printf("  - RLN witness created successfully\n");

    printf("\nCreating partial witness from witness fields\n");
    CFr *witness_identity_secret = ffi_rln_witness_input_get_identity_secret(&witness);
    CFr *witness_user_message_limit = ffi_rln_witness_input_get_user_message_limit(&witness);
    Vec_CFr witness_path_elements = ffi_rln_witness_input_get_path_elements(&witness);
    Vec_uint8 witness_path_index = ffi_rln_witness_input_get_identity_path_index(&witness);
    PartialWitnessResult partial_witness_result =
        ffi_rln_partial_witness_input_new(witness_identity_secret, witness_user_message_limit,
                                             &witness_path_elements, &witness_path_index);
    ffi_cfr_free(witness_identity_secret);
    ffi_cfr_free(witness_user_message_limit);
    ffi_vec_cfr_free(witness_path_elements);
    ffi_vec_u8_free(witness_path_index);
    if (!partial_witness_result.ok)
    {
        fprintf(stderr, "Partial witness creation error: %s\n", partial_witness_result.err.ptr);
        ffi_c_string_free(partial_witness_result.err);
        return EXIT_FAILURE;
    }
    PartialWitness *partial_witness = partial_witness_result.ok;
    printf("  - partial witness created successfully\n");

    printf("\nGenerating partial ZK proof\n");
    PartialProofResult partial_proof_result =
        ffi_rln_generate_partial_proof(&rln_instance, &partial_witness);
    if (!partial_proof_result.ok)
    {
        fprintf(stderr, "Partial proof generation error: %s\n", partial_proof_result.err.ptr);
        ffi_c_string_free(partial_proof_result.err);
        return EXIT_FAILURE;
    }
    PartialProof *partial_proof = partial_proof_result.ok;
    printf("  - partial proof generated successfully\n");

    printf("\nFinishing proof with full witness\n");
    ProofResult full_proof_result =
        ffi_rln_finish_proof(&rln_instance, &partial_proof, &witness);
    if (!full_proof_result.ok)
    {
        fprintf(stderr, "Finish proof error: %s\n", full_proof_result.err.ptr);
        ffi_c_string_free(full_proof_result.err);
        return EXIT_FAILURE;
    }
    Proof *full_proof = full_proof_result.ok;
    printf("  - partial proof finished successfully\n");

    printf("\nVerifying full proof\n");
    CBoolResult verify_full_result = verify_stateful_proof(&rln_instance, &full_proof, x);
    if (verify_full_result.err.ptr)
    {
        fprintf(stderr, "Full proof verification error: %s\n", verify_full_result.err.ptr);
        ffi_c_string_free(verify_full_result.err);
        return EXIT_FAILURE;
    }
    if (verify_full_result.ok)
    {
        printf("  - full proof verified successfully\n");
    }
    else
    {
        printf("Full proof verification failed\n");
        return EXIT_FAILURE;
    }

    ffi_rln_proof_free(full_proof);
    ffi_rln_partial_proof_free(partial_proof);
    ffi_rln_partial_witness_input_free(partial_witness);
    ffi_rln_witness_input_free(witness);
    ffi_cfr_free(message_id);
    ffi_cfr_free(x);
    ffi_cfr_free(external_nullifier);
    ffi_rln_merkle_proof_free(merkle_proof);
    member_free(&member);
    ffi_rln_free(rln_instance);
    return EXIT_SUCCESS;
}
